### ISSUE TITLE
Reactive DB support

### ISSUE THREAD

**Issue by jbaxleyiii** [meteor/data#6]
@stubailo we've been working on a GQL server for our meteor apps and are pretty close to launching. We are now starting the reactive query layer as you described in the transmission episode and are going to be supporting MongoDB, MySQL, MSSQL (and webhooks for reactive apis) as they fit our current stack ,and using redis as a caching mechanism.

Do you have any thoughts on if / how the actual queries would be adjusted using what you are working on and if there are any reserved arguments (like `reactive: bool`) that we can go ahead an implement around so we can be useful in prototyping alongside you guys?

We currently write our sql queries in .sql files and use node to parse and inject variables that way we can have meaningful git history and pretty much copy/paste into our db query tooling for quick iteration. I don't know if this would be seen as an anti-pattern or not. Just wanted to mention it.

Our biggest boon to the reactive work (a lot of our currently supported mappings don't *have* to be reactive) is for cache breaking. Luckily I think they work hand in hand. In all honesty, we are roughing out having the caching layer (default to redis with in memory / file backup) to be what powers the reactivity on queries as well. So the db change would actually invalidate the cache which would update the queries. This may be the wrong direction?

Looking forward to your thoughts!

**Comment by stubailo** [comment]
Wow, this is awesome! I'm glad that you are working in a similar direction, let's team up and get on the same page!

> if there are any reserved arguments (like reactive: bool) that we can go ahead an implement around so we can be useful in prototyping alongside you guys?

We are hoping that the GraphQL team will add support for query annotations soon, so that we can use that and avoid taking up arguments. So I wouldn't worry about that and just use whatever arguments make sense for your case, I'm sure we can converge later once we both get something working!

> We currently write our sql queries in .sql files and use node to parse and inject variables that way we can have meaningful git history and pretty much copy/paste into our db query tooling for quick iteration. I don't know if this would be seen as an anti-pattern or not. Just wanted to mention it.

I'm curious (this is something we aren't super sure about) - what is your strategy for working with super complex SQL queries and GraphQL? GraphQL seems optimized for having individual objects, rather than having a big query that pulls out a lot of data at once, do you guys have some ideas about how to make that work?

> Our biggest boon to the reactive work (a lot of our currently supported mappings don't have to be reactive) is for cache breaking. Luckily I think they work hand in hand. In all honesty, we are roughing out having the caching layer (default to redis with in memory / file backup) to be what powers the reactivity on queries as well. So the db change would actually invalidate the cache which would update the queries. This may be the wrong direction?

I think this is pretty similar to what we are thinking, except we have been hoping to store as little data in the reactive cache thing as possible. But it is totally true and great that the same mechanism can be used for cache invalidation and reactivity (if you squint enough, reactivity is just invalidating the cache on the client!) I guess we were looking at caching the actual data as a nice-to-have feature, not part of the V1, but if it's very simple to implement perhaps it could be part of the initial version as well.

> we've been working on a GQL server for our meteor apps and are pretty close to launching. 

Really excited to see what you have, I think @justinsb and I would be very interested in working together to see what your use cases are, what you have implemented so far, and how it's working for you. Is any of the work open source, or is it mostly internal to your application?

**Comment by jbaxleyiii** [comment]
Query annotations seem like the right direction to me too. I'm great with that being the plan for implementation.

> I'm curious (this is something we aren't super sure about) - what is your strategy for working with super complex SQL queries and GraphQL?

Write now our sql binding are almost a 100% direct copy paste with a little hacking on top from a previous [mysql -> mongo sync tool](https://github.com/NewSpring/heighliner-expression-engine) so they are too large right now. We are planning on refactoring into much smaller and more reusable queries much like [this](https://github.com/NewSpring/Heighliner/blob/master/ee/util/images.sql) which only pulls images from our CMS's mysql db when given the appropriate data. Still from a pure dev interface, the `.sql` files are great to write and skim if you are coming from a sql background like a few of our team do.

We are also using an REST api with oData to pull from a [.net based management system](https://github.com/NewSpring/Rock) but plan on refactoring that to direct MSSQL access (probably using [squelize](https://www.npmjs.com/package/sequelize))

> I guess we were looking at caching the actual data as a nice-to-have feature, not part of the V1

This is more due to our current stack (and the API requests instead of direct db access) but having Redis as a data caching layer has been fantastic. We load tested our meteor app yesterday and today that is pulling from the GQL app and were really happy with the results, mainly due to caching as many nodes as possible. You can really see the effects here (caching deploy done at 10:10)

![cache](https://github.com/NewSpring/load-testing/raw/master/results/apollos-02-02-2016/img/cache.png)

And because of that subsequent load test only pushed our response time up to 40ms

![cached](https://github.com/NewSpring/load-testing/raw/master/results/apollos-02-02-2016/img/cache_good.png)

So for us its a needed part and why we think it could serve the two purposes well.

> Is any of the work open source, or is it mostly internal to your application?

All of it is :+1: Its certainly specific to our application but once we get to a stable point are planning on refactoring out as much as possible.

[Apollos](https://github.com/NewSpring/Apollos/tree/card-ui) is the shared core that serves three applications right now

[NewSpring App](https://github.com/NewSpring/newspring-app) is our cordova app for iOS, Android, and hopefully Win 10 soon and [newwwspring.cc](https://github.com/NewSpring/newwwspring.cc) is the future of our front facing site. We hope to eventually merge them but it was too large to do together at start.

They all are / are going to be pulling data from [Heighliner](https://github.com/NewSpring/Heighliner) which is our graph app. It was built in under a week so please forgive how disjointed it looks!

We have a CMS using mysql (expression engine, soon to be moved to CraftCMS), and an RMS using mssql (Rock), as well as our app database in Mongo for user centric settings / etc related just to the app. We are working to add Neo4j and AWS machine learning / lamda into the mix so GraphQL was a great choice for us. All in all, I couldn't be happier with this direction.

**Comment by ramsaylanier** [comment]
I'll chime in and say that using Sequelize with GraphQL has been very very awesome. I'm using it on [this project](https://github.com/ramsaylanier/WordPressExpress) with a WordPress MYSQL database. 

I don't have much SQL experience, but working with promise chaining from Sequelize has been great.

**Comment by jbaxleyiii** [comment]
@ramsaylanier yeah it looks fantastic. https://github.com/ramsaylanier/wordexpress-schema/blob/master/src/db.js#L134-L145 especially

**Comment by ramsaylanier** [comment]
@jbaxleyiii yeah! It took me a while, because the way WordPress handles post thumbnails is kinda crazy, but now you can just ask for the thumbnail with a post and automatically get it. Very powerful.

I've done some work with CraftCMS too, and the same thing goes.
